src/Microsoft.OData.Edm/Csdl/Parsing/SchemaJsonParser.cs

```diff
@@ -0,0 +1,1564 @@
+            {
+                switch (propertyName)
+                {
+                    case "$Kind":
+                        // skip this verification, because it's verified.
+                        break;
+
+                    case "$BaseType":
+                        // The value of $BaseType is the qualified name of the base type.
+                        baseType = propertyValue.ParseAsString(context);
+                        break;
+
+                    case "$Abstract":
+                        // The value of $Abstract is one of the Boolean literals true or false. Absence of the member means false.
+                        isAbstract = propertyValue.ParseAsBoolean(context);
+                        break;
+
+                    case "$OpenType":
+                        // The value of $OpenType is one of the Boolean literals true or false. Absence of the member means false.
+                        isOpen = propertyValue.ParseAsBoolean(context);
+                        break;
+
+                    default:
+                        CsdlElement element = TryParseStructuredTypeElement(propertyName, propertyValue, context);
+                        if (element != null)
+                        {
+                            members.Add(element);
+                        }
+                        else
+                        {
+                            ReportUnexpectedElement(propertyValue, context);
+                        }
+
+                        break;
+                }
+            });
+
+            CsdlComplexType csdlComplexType = new CsdlComplexType(name,
+                baseType,
+                isAbstract == null ? false : isAbstract.Value, // default is false
+                isOpen == null ? false : isOpen.Value, // default is false
+                members.OfType<CsdlProperty>(),
+                members.OfType<CsdlNavigationProperty>(),
+                context.Location());
+
+            members.OfType<CsdlAnnotation>().ForEach(a => csdlComplexType.AddAnnotation(a));
+            return csdlComplexType;
+        }
+
+        /// <summary>
+        /// Parse the <see cref="JsonElement"/> to <see cref="CsdlKey"/>.
+        /// </summary>
+        /// <param name="name">The name.</param>
+        /// <param name="jsonValue">The json value to parse from.</param>
+        /// <param name="jsonPath">The json path of this value.</param>
+        /// <returns>the parsed <see cref="CsdlKey"/>.</returns>
+        internal static CsdlKey ParseCsdlKey(string name, JsonElement keyArray, JsonParserContext context)
+        {
+            // The value of $Key is an array with one item per key property.
+            Debug.Assert(keyArray.ValueKind == JsonValueKind.Array);
+            Debug.Assert(name == "$Key", "The name should be $Key.");
+
+            IList<CsdlPropertyReference> properties = keyArray.ParseAsArray(context, (v, p) =>
+            {
+                if (v.ValueKind == JsonValueKind.Object)

```

Comment:
> JsonValueKind.Objec [](start = 34, length = 20)

Is this logic correct? Would "Info/ID" report a ValueKind of Object?  Wouldn't it just be a string, and this code never get exercised?

Reply:
I think it's correct. the value of $Key is an array.
So, I use "ParseAsArray" to parse each item in "keyArray" (Line 655).

the item type in this $Key array is varied. 
1) if It's a normal key item, the item is a string. 
2) If it's a key alias, the item is an object.

Here's an example of #1.
```Json
"$Key": [
      "ID1",
      "ID2"
]
```
Here's an example of #2.
```Json
"$Key": [
      { "Key1Alias" : "Info/ID1" }
      { "Key2Alias" : "Info/ID2" }
]
```
     



---
In reply to: [443077448](https://github.com/OData/odata.net/pull/1775#discussion_r443077448) [](ancestors = 443077448)